data_preparation/predictor_data_prep.r

```diff
@@ -245,205 +314,142 @@ process_indicators <- function(crop, region, start_year, end_year, crop_mask_fil
       crop_mask_vals$ID <- NULL
       names(crop_mask_vals) <- c("crop_area_fraction", "x", "y", "adm_id")
 
-      ind_df <- as.data.frame(cbind(ind_vals, crop_mask_vals$crop_area_fraction))
-      ind_df <- na.exclude(ind_df)
-      names(ind_df)[names(ind_df) == "crop_mask_vals$crop_area_fraction"] <- "crop_area_fraction"
-      colnames(ind_df) <- c("indicator", "x",	"y", "adm_id",	"crop_area_fraction")
+      result <- as.data.frame(cbind(indicator = ind_vals, crop_area_fraction = crop_mask_vals$crop_area_fraction))
+      result <- na.exclude(result)
 
       # for now, "drainage_class" is the only categorical data
-      if (is_cat){
+      if (is_categorical){
         # aggregate area fraction by category
-        ind_df <- aggregate(list(sum_weight=ind_df$crop_area_fraction),
-                            by=list(adm_id=ind_df$adm_id,
-                                    indicator=ind_df$indicator),
+        result <- aggregate(list(sum_weight=result$crop_area_fraction),
+                            by=list(adm_id=result$adm_id,
+                                    indicator=result$indicator),
                             FUN=sum)
         # order by sum_weight
-        # print(head(ind_df))
-        ind_df <- ind_df[order(ind_df$adm_id, -ind_df$sum_weight),]
+        # print(head(result))
+        order(result$adm_id, result$sum_weight, decreasing=TRUE)
         # keep only the first item per adm_id
-        ind_df <- ind_df[ !duplicated(ind_df$adm_id), ]
-        ind_df$sum_weight <- NULL
+        result <- result[ !duplicated(result$adm_id), ]
+        result$sum_weight <- NULL
       } else {
         # aggregate
-        ind_df$weighted_ind <- ind_df$indicator * ind_df$crop_area_fraction
-        ind_df <- aggregate(list(sum_ind=ind_df$weighted_ind,
-                                 sum_weight=ind_df$crop_area_fraction),
-                            by=list(adm_id=ind_df$adm_id), FUN=sum)
-        ind_df$indicator <- ind_df$sum_ind/ind_df$sum_weight
-        ind_df <- ind_df[, c("adm_id", "indicator")]
+        result$weighted_ind <- result$indicator * result$crop_area_fraction
+        result <- aggregate(list(sum_ind=result$weighted_ind,
+                                 sum_weight=result$crop_area_fraction),
+                            by=list(adm_id=result$adm_id), FUN=sum)
+        result$indicator <- result$sum_ind/result$sum_weight
```

Comment:
I will need an example. Thanks.

Reply:
Actually, it would work easier in one step with a custom function, rather than with two extracts. First, put together the categorical raster and the crop fraction into a single raster, then do something like:
```r
Mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

GetWeights = function(layers)
{
  DominantClass = Mode(layers[[1]]) # Assuming the categorical raster is the first layer
  return(sum(layers[DominantClass])) # Assuming the layer number matches the class number, might need to do some extra math
}

extract(categorical, shapes, fun=GetWeights)
```

Reply:
This is tracked by #306